snapcraft/tests/test_plugin_autotools.py

```diff
@@ -0,0 +1,155 @@
+
+        configflags_default = configflags['default']
+        self.assertEqual(configflags_default, [],
+                         'Expected "configflags" "default" to be [], but '
+                         'it was {}'.format(configflags_default))
+
+        configflags_items = configflags['items']
+        self.assertTrue('type' in configflags_items,
+                        'Expected "type" to be included in "configflags" '
+                        '"items"')
+
+        configflags_items_type = configflags_items['type']
+        self.assertEqual(configflags_items_type, 'string',
+                         'Expected "configflags" "items" "type" to be '
+                         '"string", but it was "{}"'
+                         .format(configflags_items_type))
+
+    @mock.patch.object(autotools.AutotoolsPlugin, 'run')
+    def test_build_configure(self, run_mock):
+        plugin = autotools.AutotoolsPlugin('test-part', self.options)
+        os.makedirs(plugin.sourcedir)
+
+        # Create both configure and autogen.sh.
+        # Configure should take precedence.
+        open(os.path.join(plugin.sourcedir, 'configure'), 'w').close()
+        open(os.path.join(plugin.sourcedir, 'autogen.sh'), 'w').close()
+
+        plugin.build()
+
+        self.assertEqual(2, run_mock.call_count)
+        run_mock.assert_has_calls([
+            mock.call(['./configure', '--prefix=']),
+            mock.call(['make', 'install',
+                       'DESTDIR={}'.format(plugin.installdir)])
+        ])
+
+    @mock.patch.object(autotools.AutotoolsPlugin, 'run')
+    def test_build_autogen(self, run_mock):
+        plugin = autotools.AutotoolsPlugin('test-part', self.options)
+        os.makedirs(plugin.sourcedir)
+
+        # No configure-- only autogen.sh. Make sure it's executable.
+        open(os.path.join(plugin.sourcedir, 'autogen.sh'), 'w').close()
+        os.chmod(os.path.join(plugin.sourcedir, 'autogen.sh'),
+                 stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR)
+
+        plugin.build()
+
+        self.assertEqual(3, run_mock.call_count)
+        run_mock.assert_has_calls([
+            mock.call(['env', 'NOCONFIGURE=1', './autogen.sh']),
+            mock.call(['./configure', '--prefix=']),
+            mock.call(['make', 'install',
+                       'DESTDIR={}'.format(plugin.installdir)])
+        ])
+
+    @mock.patch.object(autotools.AutotoolsPlugin, 'run')
+    def test_build_autoreconf(self, run_mock):
+        plugin = autotools.AutotoolsPlugin('test-part', self.options)
+        os.makedirs(plugin.sourcedir)
+
+        # No configure or autogen.sh.
+
+        plugin.build()
+
+        self.assertEqual(3, run_mock.call_count)
+        run_mock.assert_has_calls([
+            mock.call(['autoreconf', '-i']),
+            mock.call(['./configure', '--prefix=']),
+            mock.call(['make', 'install',
+                       'DESTDIR={}'.format(plugin.installdir)])
+        ])
+
+    @mock.patch('os.chmod')
```

Comment:
why are you mocking `os.chmod` but actually creating the file?

It seems less error prone to not mock `os.chmod` and `os.stat` it